In [1]:
import pandas as pd
import numpy as np
df1= pd.read_csv('Data/superhero_info - superhero_info.csv')
df2= pd.read_csv('Data/superhero_powers - superhero_powers.csv')

In [2]:
df1.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [3]:
df2.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [4]:
df1[['Hero','Publisher']] = df1['Hero|Publisher'].str.split('|',expand=True)
df1.head()
df1=df1.drop(columns=['Hero|Publisher'])

In [5]:
HW = df1.loc[0,"Measurements"]
print(type(HW))
HW
import json
HW = HW.replace("'",'"')
HW

<class 'str'>


'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [6]:
fixed_HW = json.loads(HW)
print(type(fixed_HW))
fixed_HW

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [7]:
df1['Measurements'] = df1['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
df1['Measurements'] = df1['Measurements'].apply(json.loads)
df1['Measurements'].head()
test_HW = df1.loc[0, 'Measurements']
print(type(test_HW))
test_HW

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [8]:
HWLong = df1['Measurements'].apply(pd.Series)
HWLong
df1 = pd.concat((df1, HWLong), axis = 1)
df1.head(5)

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg  
2          DC Comics  185.0 cm   90.0 kg  
3      Marvel Comics  203.0 cm  441.0 kg  
4      Marvel Comics  193.0 cm  122.0 kg

In [9]:
df1=df1.drop(columns=['Measurements'])
df1['Height'] = df1['Height'].str.replace(' cm', '')
df1['Weight'] = df1['Weight'].str.replace(' kg', '')

In [10]:
df1['Height'] = df1['Height'].astype(float)
df1['Weight'] = df1['Weight'].astype(float)

In [11]:
df1.head()
df1 = df1[['Hero', 'Publisher', 'Gender', 'Eye color', 'Race', 'Hair color', 'Height', 'Skin color', 'Alignment', 'Weight']]

In [12]:
df2.loc[2,'Powers']

#df2['Powers_split'] = df2['Power'].str.replace("'",'"')
#df2['Powers'].dtypes


'Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner'

In [13]:
df2['Powers']

df2['PowersList'] = df2['Powers'].apply(lambda x: x.split(','))

In [16]:
#df2['Powers'] = df2['Powers'].apply(json.loads)
# check results
df2['PowersList'].head()
#df2['PowersList'] = df2['PowersList'].apply(lambda x: x.split(','))

0      [Agility, Super Strength, Stamina, Super Speed]
1    [Accelerated Healing, Durability, Longevity, S...
2    [Agility, Accelerated Healing, Cold Resistance...
3                                 [Lantern Power Ring]
4    [Accelerated Healing, Intelligence, Super Stre...
Name: PowersList, dtype: object

In [18]:
exploded = df2.explode('PowersList')
exploded

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
..         ...                                                ...   
665    Zatanna  Cryokinesis,Telepathy,Magic,Fire Control,Proba...   
666       Zoom  Super Speed,Intangibility,Time Travel,Time Man...   
666       Zoom  Super Speed,Intangibility,Time Travel,Time Man...   
666       Zoom  Super Speed,Intangibility,Time Travel,Time Man...   
666       Zoom  Super Speed,Intangibility,Time Travel,Time Man...   

              PowersList  
0                Agility  
0         Super Strength  
0                Stamina  
0            Super Speed  
1    Accelerated Healing  
..                   ...  
665      Weather Control  
666          Super Speed  
666        Intangibility  
666          Time Travel  
666    Time Manipulation  

[5874 rows x 3 columns]

In [19]:
cols_to_make = exploded['PowersList'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [22]:
for col in cols_to_make:
    df2[col] = df2['Powers'].str.contains(col)
    

In [23]:
df2

hero_names                                             Powers  \
0            3-D Man         Agility,Super Strength,Stamina,Super Speed   
1             A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2         Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3           Abin Sur                                 Lantern Power Ring   
4        Abomination  Accelerated Healing,Intelligence,Super Strengt...   
..               ...                                                ...   
662  Yellowjacket II                 Flight,Energy Blasts,Size Changing   
663             Ymir  Cold Resistance,Durability,Longevity,Super Str...   
664             Yoda  Agility,Stealth,Danger Sense,Marksmanship,Weap...   
665          Zatanna  Cryokinesis,Telepathy,Magic,Fire Control,Proba...   
666             Zoom  Super Speed,Intangibility,Time Travel,Time Man...   

                                            PowersList  Agility  \
0      [Agility, Super Strength, Stamina, Super Speed]     True   
1    [Accelerated Healing, Durability, Longevity, S...    False   
2    [Agility, Accelerated Healing, Cold Resistance...     True   
3                                 [Lantern Power Ring]    False   
4    [Accelerated Healing, Intelligence, Super Stre...    False   
..                                                 ...      ...   
662             [Flight, Energy Blasts, Size Changing]    False   
663  [Cold Resistance, Durability, Longevity, Super...    False   
664  [Agility, Stealth, Danger Sense, Marksmanship,...     True   
665  [Cryokinesis, Telepathy, Magic, Fire Control, ...    False   
666  [Super Speed, Intangibility, Time Travel, Time...    False   

     Super Strength  Stamina  Super Speed  Accelerated Healing  Durability  \
0              True     True         True                False       False   
1              True     True        False                 True        True   
2              True     True        False                 True        True   
3             False    False        False                False       False   
4              True     True         True                 True       False   
..              ...      ...          ...                  ...         ...   
662           False    False        False                False       False   
663            True    False        False                False        True   
664           False     True         True                False       False   
665           False    False        False                False       False   
666           False    False         True                False       False   

     Longevity  ...  Weather Control  Omnipresent  Omniscient  \
0        False  ...            False        False       False   
1         True  ...            False        False       False   
2         True  ...            False        False       False   
3        False  ...            False        False       False   
4        False  ...            False        False       False   
..         ...  ...              ...          ...         ...   
662      False  ...            False        False       False   
663       True  ...            False        False       False   
664       True  ...            False        False       False   
665      False  ...             True        False       False   
666      False  ...            False        False       False   

     Hair Manipulation  Nova Force  Odin Force  Phoenix Force  \
0                False       False       False          False   
1                False       False       False          False   
2                False       False       False          False   
3                False       False       False          False   
4                False       False       False          False   
..                 ...         ...         ...            ...   
662              False       False       False          False   
663              False       False       False          False   
6

In [24]:
df2=df2.drop(columns=['Powers'])
df2 = df2.rename(columns={'hero_names': 'Hero'})


In [25]:
df3=pd.merge(df1,df2, on='Hero')
df3

Hero          Publisher  Gender Eye color               Race  \
0             A-Bomb      Marvel Comics    Male    yellow              Human   
1         Abe Sapien  Dark Horse Comics    Male      blue      Icthyo Sapien   
2           Abin Sur          DC Comics    Male      blue            Ungaran   
3        Abomination      Marvel Comics    Male     green  Human / Radiation   
4      Absorbing Man      Marvel Comics    Male      blue              Human   
..               ...                ...     ...       ...                ...   
458     Yellowjacket      Marvel Comics    Male      blue              Human   
459  Yellowjacket II      Marvel Comics  Female      blue              Human   
460             Yoda       George Lucas    Male     brown     Yoda's species   
461          Zatanna          DC Comics  Female      blue              Human   
462             Zoom          DC Comics    Male       red            Unknown   

           Hair color  Height Skin color Alignment  Weight  ...  \
0             No Hair   203.0    Unknown      good   441.0  ...   
1             No Hair   191.0       blue      good    65.0  ...   
2             No Hair   185.0        red      good    90.0  ...   
3             No Hair   203.0    Unknown       bad   441.0  ...   
4             No Hair   193.0    Unknown       bad   122.0  ...   
..                ...     ...        ...       ...     ...  ...   
458             Blond   183.0    Unknown      good    83.0  ...   
459  Strawberry Blond   165.0    Unknown      good    52.0  ...   
460             White    66.0      green      good    17.0  ...   
461             Black   170.0    Unknown      good    57.0  ...   
462             Brown   185.0    Unknown       bad    81.0  ...   

    Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0             False        False       False              False       False   
1             False        False       False              False       False   
2             False        False       False              False       False   
3             False        False       False              False       False   
4             False        False       False              False       False   
..              ...          ...         ...                ...         ...   
458           False        False       False              False       False   
459           False        False       False              False       False   
460           False        False       False              False       False   
461            True        False       False              False       False   
462           False        False       False              False       False   

     Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0         False          False               False    False           False  
1         False          False               False    False           False  
2         False          False               False    False           False  
3         False          False               False    False           False  
4         False          False               False    False           False  
..          ...            ...                 ...      ...             ...  
458       False          False               False    False           False  
459       False          False               False    False           False  
460       False          False               False    False           False  
461       False          False               False    False           False  
462       False          False               False    False           False  

[463 rows x 178 columns]

In [28]:
average_speed_weight = df3.groupby(df3['Super Speed'])['Weight'].mean()
average_speed_weight

Super Speed
False    101.773585
True     129.404040
Name: Weight, dtype: float64

What is the average height of heroes for each publisher?


Average weight of superheroes with super speed is 129.40 kg.
Average weight of superheroes without super speed is 101.77 kg.

In [29]:
average_publisher_height = df3.groupby(df3['Publisher'])['Height'].mean()
average_publisher_height

Publisher
DC Comics            181.923913
Dark Horse Comics    176.909091
George Lucas         159.600000
Image Comics         211.000000
Marvel Comics        191.546128
Shueisha             171.500000
Star Trek            181.500000
Team Epic TV         180.750000
Unknown              178.000000
Name: Height, dtype: float64

The average height for each publisher is listed above.
